### PR TITLE
make HTTP routes async

### PR DESCRIPTION
**Pull request by technillogue**
this doesn't change any logic, but as far as I understand sync routes use threads instead of the starlette event loop: https://fastapi.tiangolo.com/async/#path-operation-functions

### CODE REVIEW

**Review by nickstenning — COMMENTED** [review]
I believe this may not be as harmless as it looks.

There's a `--threads` argument to `cog.server.http` and I believe that it functions as a concurrency control for the synchronous prediction create endpoint.

I believe at least some people rely on this behaviour so they can use the network socket's accept queue as a mechanism to queue up work. If this breaks that behaviour (i.e. those requests are immediately accepted and we serve a 409 instead) that may break those folks' expectations.

Is this change a requirement or just a nice-to-have for the batching/concurrent predictions work?

**Comment by technillogue** [comment]
Do we have any specific examples of someone actually using that, or is hypothetical?

I _think_ it will be required somewhere down the line, I put it up front here because it can be reviewed as an independent PR

**Comment by nickstenning** [comment]
Yes, in that if I remember correctly, we broke it previously while working on Cog and had to put it back.

**Comment by technillogue** [comment]
we could stick a timeout or something where the endpoint waits for the worker for a similar time as what the tcp timeout would have been? do you remember what exactly the use case was and why they didn't want to use retries instead?

File: python/cog/server/http.py [hunk 1]
```diff
@@ -79,9 +81,20 @@ def create_app(
...
+            async with http_semaphore:
+                return await f(*args, **kwargs)
+
+        return wrapped
+
     @app.on_event("startup")
     def startup() -> None:
         # https://github.com/tiangolo/fastapi/issues/4221
+        # potentially no longer necessary?
         RunVar("_default_thread_limiter").set(CapacityLimiter(threads))  # type: ignore
```
Comment (nickstenning) [inline python/cog/server/http.py:98]: I think it's definitely no longer necessary if we've verified that the semaphore does the same job.

**Review by nickstenning — APPROVED** [review]
I think it would be nice if we could have a test for this behaviour, but I'm not going to block this on that. I've tested it locally and it works well!

**Pull request merged**